[JSC] add support for `Promise.withResolvers`

dcrousso opened the pull request:
#### 0e3ae70dd28cc89fc67a92312808f26ec35c60ff
<pre>
[JSC] add support for `Promise.withResolvers`
<a href="https://bugs.webkit.org/show_bug.cgi?id=259153">https://bugs.webkit.org/show_bug.cgi?id=259153</a>

Reviewed by Yusuke Suzuki.

This allows for developers to `resolve`/`reject` a `Promise` after it&apos;s been created instead of having to provide a function to `new Promise`.

For example, the following
```
function foo(target) {
    return new Promise((resolve, reject) =&gt; {
        target.addEventListener(&quot;good&quot;, resolve);
        target.addEventListener(&quot;bad&quot;, reject);
    });
}
```
can now be rewritten as
```
function foo(target, event) {
    let {promise, resolve, reject} = Promise.withResolvers();

    target.addEventListener(&quot;good&quot;, resolve);
    target.addEventListener(&quot;bad&quot;, reject);

    return promise;
}
```

Spec: <a href="https://tc39.es/proposal-promise-with-resolvers/">https://tc39.es/proposal-promise-with-resolvers/</a>
Proposal: <a href="https://github.com/tc39/proposal-promise-with-resolvers">https://github.com/tc39/proposal-promise-with-resolvers</a>

* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(withResolvers): Added.
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSPromiseConstructor::finishCreation):

* Source/JavaScriptCore/runtime/OptionsList.h:
Add an off-by-default feature flag for this.

* JSTests/stress/promise-withResolvers.js: Added.

Canonical link: <a href="https://commits.webkit.org/266015@main">https://commits.webkit.org/266015@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3e944b3b882b4d507c82bcb20347b30d71a4003

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12550 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12879 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14290 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12038 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15381 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12897 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14726 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12714 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13495 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10615 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14716 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10773 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11358 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18451 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/10672 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11857 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11523 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14721 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/11874 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12007 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9926 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/12607 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11247 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3300 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3083 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15580 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/12963 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11863 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3090 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -80,6 +80,11 @@ void JSPromiseConstructor::finishCreation(VM& vm, JSPromisePrototype* promisePro
     putDirectWithoutTransition(vm, vm.propertyNames->length, jsNumber(1), PropertyAttribute::DontEnum | PropertyAttribute::ReadOnly);
     putDirectNonIndexAccessorWithoutTransition(vm, vm.propertyNames->speciesSymbol, speciesSymbol, PropertyAttribute::Accessor | PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum);
+
+    JSGlobalObject* globalObject = promisePrototype->globalObject();
```
Why not using `globalObject()` of this `JSPromiseConstructor`?

Constellation commented on Source/JavaScriptCore/builtins/PromiseConstructor.js:
```diff
@@ -368,6 +368,19 @@ function resolve(value)
+        promise: promiseCapability.@promise,
+        resolve: promiseCapability.@resolve,
+        reject: promiseCapability.@reject,
+    };
```
I think we should replace all of `@promise` with `promise`, `@resolve` with `resolve`, `@reject` with `reject` and making this function as follows (there is no reason to continue using `@promise` etc.). Currently, `newPromiseCapability` is allocating unnecessary object here.

```
function withResolvers()
{
    "use strict";
    return @newPromiseCapability(this);
}
```

Constellation replied:
But keep in mind that `newPromiseCapability`'s function name itself must be `@resolve` etc. to avoid naming them as `"resolve"`.

dcrousso replied:
are `@promise`/`@resolve`/`@reject` referenceable outside of builtins?  i tried doing as you suggested (i.e. just `return @newPromiseCapability(this);`) and the resulting object had no properties on it 🤔

or are you suggesting that we rework `newPromiseCapability` itself to return `promise`/`resolve`/`reject` instead (and updating all the callsites to reflect that too)?

kmiller68 replied:
He's suggesting the latter. To return `promise/resolve/reject` and rework the other callsites.

dcrousso replied:
would it be OK if i did that in a followup?  AFAICT looking at it right now it seems like a _lot_ of places to update (as well as other things to change, such as replacing `vm.propertyNames->promisePrivateName()` with something else)

dcrousso replied:
to elaborate, id rather this PR just be about adding this new feature and have a separate PR for optimizing it (since AFAICT that optimization involves changing a _lot_ of other things, which probabilistically could have more issues since it's a larger change)

also, is there concern about `newPromiseCapability` no longer matching the spec in the sense that we're using plain properties `promise`/`resolve`/`reject` instead of internal slots `[[Promise]]`/`[[Resolve]]`/`[[Reject]]` (unless im mistaken in thinking that our `@promise`/`@resolve`/`@reject` is equivalent to that)?

dcrousso replied:
https://github.com/WebKit/WebKit/pull/15808

Constellation reviewed (approved):
r=me, but `@resolve` refactoring is mandatory requirement for enabling this feature.

dcrousso committed:
[JSC] add support for `Promise.withResolvers`
https://bugs.webkit.org/show_bug.cgi?id=259153

Reviewed by Yusuke Suzuki.

This allows for developers to `resolve`/`reject` a `Promise` after it's been created instead of having to provide a function to `new Promise`.

For example, the following
```
function foo(target) {
    return new Promise((resolve, reject) => {
        target.addEventListener("good", resolve);
        target.addEventListener("bad", reject);
    });
}
```
can now be rewritten as
```
function foo(target, event) {
    let {promise, resolve, reject} = Promise.withResolvers();

    target.addEventListener("good", resolve);
    target.addEventListener("bad", reject);

    return promise;
}
```

Spec: https://tc39.es/proposal-promise-with-resolvers/
Proposal: https://github.com/tc39/proposal-promise-with-resolvers

* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(withResolvers): Added.
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSPromiseConstructor::finishCreation):

* Source/JavaScriptCore/runtime/OptionsList.h:
Add an off-by-default feature flag for this.

* JSTests/stress/promise-withResolvers.js: Added.

Canonical link: https://commits.webkit.org/266015@main
